### DISCUSSION TITLE
How to migragte plugin "parcel-plugin-transcrypt" from V1 to V2

### DISCUSSION THREAD

**Discussion by aihuasxy** [parcel-bundler/parcel#7423] [General] [answered]
I want to migrate plugin [parcel-plugin-transcrypt](https://www.npmjs.com/package/parcel-plugin-transcrypt) to V2, have some problems.

parcel-plugin-transcrypt's code is [here](https://github.com/QQuick/Transcrypt/blob/merge_parcel_bundler2/parcel/asset.js)

it's process is described as follows：

Step1 Call transcrypt to compile the python code into JS files and Map files. All files are placed in the \__target__ directory. After transcrypt excuted, a file named asset.project will be generated, which contains the source file list and the map file list.

for example
`"modules": [
    {
      "source": "/home/***/__runtime__.py",
      "target": "/home/code/ch12/__target__/org.transcrypt.__runtime__.js"
    },
    {
      "source": "/home/code/ch12/app.py",
      "target": "/home/code/ch12/__target__/app.js"
    },
    {
      "source": "/home/code/ch12/pyreact.py",
      "target": "/home/code/ch12/__target__/pyreact.js"
    }
  ]
`
in async load()

Step2 Parse the project file and put the map files in the dependency list. If don’t add it, you cannot watch the changed files.

in async collectDependencies()
this.addDependency(mod.source, {includedInParent: true});

Step3 returns the result in generate

generate() {
        return [ {
            type: this.type,
            value: this.contents,
            sourceMap: this.sourceMap,
        } ];
    }

V2 plugin system is different from V1, so i don’t how to migrate
Step1 implement in transfom?
Step2 down't know where to add the dependency list
Step3 retrun in transfom?

**Comment by mischnic** [comment, marked as answer]
Would something like this work in your case (you have to provide the used functions)?

```js
export default new Transformer({
  async transform({asset, resolve}) {
	let inputCode = await asset.getCode();

	// Compile and bundle that file, return resulting JS and files that are used
	let result = await compilePyToJs(asset.filePath, inputCode);

	for(let f of result.includedFile) {
		asset.invalidateOnFileChange(f); // rerun this transformer after that file changed
	}

	asset.type = "js";
	asset.setCode(result.code);
    return [asset];
  },
});
```

The GLSL transformer is similar: https://github.com/parcel-bundler/parcel/blob/v2/packages/transformers/glsl/src/GLSLTransformer.js

**Reply by JennaSys** [reply]
I'm interested in implementing this as well.  Thanks for the direction, I'll see if I can get something working using this as a template.

**Reply by JennaSys** [reply]
Following up on this, I just published a [Python transformer for Parcel V2](https://www.npmjs.com/package/parcel-transformer-transcrypt) that uses Transcrypt.  I'm still working on the docs and it needs _way_ more testing in different environments than I've done so far, but I figured I'd put it out there early and get some feedback.

Feel free to [open up a GitHub issue](https://github.com/JennaSys/parcel-transformer-transcrypt/issues) if you run into any problems with it.
